[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitUserContentFilterStore.cpp

aperezdc opened the pull request:
#### de5bceefeabdceb7f0b7c7248abfef9636dc2198
<pre>
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitUserContentFilterStore.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=303869">https://bugs.webkit.org/show_bug.cgi?id=303869</a>

Reviewed by Michael Catanzaro.

* Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.cpp:
(webkit_user_content_filter_store_fetch_identifiers): Replace memory
allocation for the function result with GMallocSpan::malloc(), bringing
in bounds-checking. While at it, use checked arithmetic to calculate
the size of the allocation and handle overflows. Lastly, switch from
g_strdup() to g_strndup() to avoid an unneeded strlen() call for each
identifier copied into the result set.

Canonical link: <a href="https://commits.webkit.org/304283@main">https://commits.webkit.org/304283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21f5eef21168bea801765905bc3c1ec79d4f0656

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135021 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7432 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142530 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86857 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8054 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103173 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70417 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5715 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84026 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5533 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3144 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3125 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114743 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145228 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7109 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39780 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111548 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7161 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111908 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28414 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5369 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117332 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61053 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7157 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35497 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6930 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70732 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7162 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.cpp:
```diff
@@ -505,13 +505,17 @@ void webkit_user_content_filter_store_fetch_identifiers(WebKitUserContentFilterS
         }
-        g_task_return_pointer(task.get(), result, reinterpret_cast<GDestroyNotify>(g_strfreev));
+
+        auto result = GMallocSpan<gchar*>::zeroedMalloc(bytesToAllocate);
```
Normally I do like to zero a memory allocation, just to reduce the risk of bugs.

But in this case, since the value may be large, and since the loop guarantees we initialize all except the last element, it might be more efficient to just zero out only the final element instead?

aperezdc replied:
Sure, will do. Also I realized that we can avoid a call to `strlen()` for each copied identifier, using `g_strndup()` and passing the length of the CString instead of using `g_strdup()`.

aperezdc committed:
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitUserContentFilterStore.cpp
https://bugs.webkit.org/show_bug.cgi?id=303869

Reviewed by Michael Catanzaro.

* Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.cpp:
(webkit_user_content_filter_store_fetch_identifiers): Replace memory
allocation for the function result with GMallocSpan::malloc(), bringing
in bounds-checking. While at it, use checked arithmetic to calculate
the size of the allocation and handle overflows. Lastly, switch from
g_strdup() to g_strndup() to avoid an unneeded strlen() call for each
identifier copied into the result set.

Canonical link: https://commits.webkit.org/304283@main
